include/deal.II/base/index_set.h

```diff
@@ -79,6 +89,188 @@ class IndexSet
+     * distance is given by how many times one has to apply operator++ to the
+     * current iterator to get the argument (for a positive return value), or
+     * operator-- (for a negative return value).
+     */
+    std::ptrdiff_t operator - (const IntervalIterator &p) const;
+
+  private:
+    IntervalAccessor accessor;
+  };
+
+  /**
+   * Class that represents an iterator pointing to a single element in the IndexSet as returned by IndexSet::begin().
+   */
+  class ElementIterator
+  {
+  public:
+    /**
+     * Construct an iterator pointing to the global index @p index in the interval @p range_idx
+     */
+    ElementIterator(const IndexSet &idxset, size_type range_idx, size_type index);
+    /**
+     * Construct an iterator pointing to the end of the IndexSet.
+     */
+    ElementIterator(const IndexSet &idxset);
+
+    /**
+     * Dereferencing operator. The returned value is the index of the element inside the IndexSet.
+     */
+    const size_type &operator* () const;
+
+    /**
+     * Does this iterator point to an existing element?
+     */
+    bool is_valid () const;
+
+    /**
+     * Prefix increment.
+     */
+    ElementIterator &operator++ ();
+
+    /**
+     * Postfix increment.
+     */
+    ElementIterator operator++ (int);
+
+
+    /**
+     * Comparison.
+     */
+    bool operator == (const ElementIterator &) const;
+
+    /**
+     * Inverse of <tt>==</tt>.
+     */
+    bool operator != (const ElementIterator &) const;
+
+    /**
+     * Comparison operator.
+     */
+    bool operator < (const ElementIterator &) const;
+
+    /**
+     * Return the distance between the current iterator and the argument. The
+     * distance is given by how many times one has to apply operator++ to the
+     * current iterator to get the argument (for a positive return value), or
+     * operator-- (for a negative return value).
+     */
+    std::ptrdiff_t operator - (const ElementIterator &p) const;
+
+  private:
+    /**
+     * Advance iterator by one.
+     */
+    void advance ();
+
+    /**
+     * The parent IndexSet.
+     */
+    const IndexSet &index_set;
+    /**
+     * Index into index_set.ranges.
+     */
+    size_type range_idx;
+    /**
+     * The global index this iterator is pointing at.
+     */
+    size_type idx;
+  };
+
+  /**
+   * Return an iterator that points at the first index that is contained in this IndexSet.
+   */
+  ElementIterator begin() const;
+  /**
+   * Return an iterator that points one after the last index that is contained in this IndexSet.
+   */
+  ElementIterator end() const;
+
+  /**
+   * Return an Iterator that points at the first interval of this IndexSet.
+   */
+  IntervalIterator begin_intervals() const;
+
+  /**
+   * Return an Iterator that points one after the last interval of this IndexSet.
+   */
+  IntervalIterator end_intervals() const;
+
+  /**
    * Default constructor.
    */
   IndexSet ();
```

Comment:
Move the new functions below the constructors. We typically have the constructors come first.

You may want to consider putting all of this into a section enclosed into `/** @{ */` and `/** @} */`